text/0000-roadmap-2020.md

```diff
@@ -0,0 +1,368 @@
+The roadmap takes the form of the following goals for the project:
+
+* Prepare for a possible Rust 2021 Edition
+* Followthrough on in-progress designs and efforts
+* Improve project functioning and governance:
+  * Improve visibility into the state of initiatives and design efforts
+  * Increase mentoring, leadership, and organizational bandwidth
+  * Make design discussions more productive and less exhausting 
+
+# Motivation
+[motivation]: #motivation
+
+Every year, the Rust project plans out a roadmap, in accordance with
+[RFC 1728]. The goal of the roadmap is to
+
+* Align the Rust project on our priorities in the coming year, to help
+  teams focus their efforts on addressing the most prominent problems
+* Communicate these priorities to the community and outside world
+
+To that end, the roadmap describes the general goals that we believe the
+teams ought to be pursuing. These goals were chosen based on a number of
+sources:
+
+[RFC 1728]: https://rust-lang.github.io/rfcs/1728-north-star.html
+
+* Preliminary analysis of the [2019 survey], which took place in
+  December.
+* The [many #rust2020 blog posts][rust2020] written in response to our
+  [call for blog posts].
+* The thoughts and inputs from the members of the various Rust teams.
+
+[2019 survey]: https://blog.rust-lang.org/2019/12/03/survey-launch.html
+[rust2020]: https://readrust.net/rust-2020/
+[call for blog posts]: https://blog.rust-lang.org/2019/10/29/A-call-for-blogs-2020.html
+
+The roadmap is not meant to be "exclusive" -- that is, it's not the
+case that every single thing we do must tie in some way to the
+roadmap. But we do expect that our largest efforts will be put towards
+addressing the roadmap goals.
+
+## Structure of the roadmap
+
+The roadmap this year is based around a few central themes. These goals
+are intentionally rather broad -- they are meant to be interpreted
+throughout the year by the various teams, as they make decisions about
+what to pursue.
+
+The roadmap does not contain specific technical details or
+proposals. We encourage the individual teams to post their thoughts
+about goals and ongoing projects for 2020, either in the form of
+[Inside Rust] blog posts or as [internals] threads.
+
+[Inside Rust]: https://blog.rust-lang.org/inside-rust/index.html
+[internals]: https://internals.rust-lang.org/
+
+# The goals
+
+We have laid out three 'major goals' for Rust in 2020:
+
+* Prepare for a possible Rust 2021 Edition
+* Followthrough on in-progress designs and efforts
+* Improve project functioning and governance:
+  * Improve visibility into the state of initiatives and design efforts
+  * Increase mentoring, leadership, and organizational bandwidth
+  * Make design discussions more productive and less exhausting 
+
+## Prepare for a Rust 2021 edition
+
+[Editions] were establshed as a means to help communicate the progress of
+the Rust language and provide a rallying point for overarching pieces of work.
```

Comment:
"rally point for a marketing push" mixed into other things was definitely a bad idea.

However, "3 years" is easy to understand for users, and it's already what most people think, so if there are _any_ changes that can be arranged by the end of 2020 and pushed put the door by mid to late 2021 we should **really** try to stick to something in 2021, and anything after that just plain has to wait until 2024. Same as we push rustc every 6 weeks no matter how much or how little was done that 6 weeks.

Reply:
We discussed this a bunch internally, it seemed like consensus was either to do it as:

 - once every three years, but not push super super hard for breaking lang changes (if it slips, it slips, it'll happen next time)
 - whenever we feel it is necessary, and push the button when everything is ready

(You seem to be suggesting the latter)

an observation was that with the latter it's harder to do things very predictably, as @Lokathor said a regular 3 year cadence is easy to understand. It's also nice to have a periodic longer-term thing where the rest of the world realizes the progress we've made. I think @nikomatsakis and @nrc had some really well-written thoughts on this

Reply:
> It's also nice to have a periodic longer-term thing where the rest of the world realizes the progress we've made. I think @nikomatsakis and @nrc had some really well-written thoughts on this

I don't mind such a longer-term thing, but I think it shouldn't be organized together with editions as it pushes us to stabilize features on a specific deadline, before things are and free of known issues.

Reply:
It might push us to focus more on a given issue, but if it slips it slips.

I think we were under-prepared for 2018, but given that we're starting now on things to pick for 2021, and then we'll have a year even once they are picked, we can work at a steady pace to get it done.

Reply:
It is easy to say "if it slips it slips". In my experience however (of working on the language), once you set a deadline for stabilizing a feature in advance of the feature actually being fully baked, you tend to take shortcuts to bakedness by adding hacks and overlooking bugs.

We tried to do a lot of things in 2018, too many things, and especially things which were entirely unrelated to the actual edition, which is also why it consumed a lot of time. We are better positioned now, given that the edition system, hygiene, etc. is implemented, so it is mostly a matter of specific edition changes now. However, I don't think that's a good reason to mix edition and non-edition changes together.

Reply:
@Centril you've mentioned deadlines a few times - can you elaborate on where you find the perception of these come from? Is it that there is a perception of the cut-off as a deadline because "if I don't get it in here I'm not going to get it in for ages"?

My concern with pushing out the cut-off is it unintuitively *increases* the pressure to get it in next time, because there's now no regular schedule and "who knows when the next chance will come around".

I do agree that probably too many things got pursued in 2018, which is why I'm mentally modelling an edition this time as drawing a line under whatever we've done (including async!) and bundling 'obviously required' breaking changes with it. A 'rallying point' doesn't necessarily mean we have to get a bunch of things done in pursuit of something - it can also mean just taking stock of what we've achieved.

Reply:
Hmm, what to say. There's a lot to unpack here I barely even know where to start!

To start, I'll say that I'm not entirely sure what I think the best model of editions are. Maybe it's a good topic to explore in a more depth on an internals thread or via a few blog posts. This is partly why we left this a bit open in the roadmap.

But I think that we sort of have to choosen between two basic concepts

* an edition is sort of an extra-big **release** -- that means it should be periodic, but it's also going to be less exciting
* an edition is sort of an extra-big **feature** -- that means it should be "ready when it's ready" (RWIW), and it's going to be a big deal when it's announced

Interestingly, we've largely thought of them in Rust as the first one (though perhaps with elements of the second), but other projects (like Ember) have adopted the second. I think that both models have a lot to recommend them.

I think that if we want to try and remove some of the pressure to get things done, though, we probably want to push more towards the "extra-big release" model. The idea would be to separate the "completion of a feature" with "release of the edition", unless we are able to get that feature done *well in advance of the edition*.

This is, actually, pretty much what we did for async-await, but we didn't realize it. We cut it from Rust 2018, and only made the minimal changes (reserving keywords) that we thought we would need. Then we added it asynchronously. I would imagine that if we had editions running on a regular schedule, this would be part of what we do as planning -- we front-load the "deprecate old code" part of the edition, make sure that gets done for the deadline, and then we let the "nice stuff" get added later. Since we don't generally expect deprecation to be that common or painful, that seems fine.

Let me just phrase that another way. If we make editions run on a schedule, the emphasis for big features should be to get enough of the pre-work done so that their **deprecations** are part of the edition, but not necessarily any new additions. One nice thing about this particular edition cycle is that I don't think we have a lot of big features, so we're going to be able to just focus on smalle changes.

Bah, I have a lot more to say, but having rewritten this comment 4 or 5 times already, I'm going to stop here for the moment.

Reply:
An interesting example is the linux kernel. Linus cuts a new version at arbitrary times with the explicit intention of making the version numbers somewhat meaningless:

> But I'd like to point out (yet again) that we don't do feature-based releases, and that "5.0" doesn't mean anything more than that the 4.x numbers started getting big enough that I ran out of fingers and toes. ([Linus Torvalds](https://lkml.org/lkml/2019/3/3/236))

I used to think this was odd, but now I see that this does a good job showing progress without making everybody rush to get things into a particular version. The 2018 edition was the opposite of that: we marketed it as a collection of polished features with a theme, so everyone rushed to get their features in.

If we intentionally emphasized that the only meaning of the 2021 edition is the breaking changes, perhaps there would be less of a rush to get features into it.

In other words, we would not market the edition in any way (though we would announce it widely) or give it a theme. We would just use it to make breaking changes and take the opportunity to do a retrospective of the last 3 years progress.

Reply:
> In other words, we would not market the edition in any way (though we would announce it widely) or give it a theme. We would just use it to make breaking changes and take the opportunity to do a retrospective of the last 3 years progress.

Yes, this is sort of what I was suggesting. Either we treat it like a *release* (not really marketed that much except to say "hey, here is notable stuff that happened") or as a *feature*, but not both.

I think treating it like a release is still useful, and it still makes sense to leave plenty of time for polish -- especially because polish is so often overlooked. 

I think it'd be great if we are able to establish a rhythm of "every 3 years, we do an edition change, which allows us to keep the rust language evolving as well as fixing small warts", though I do think we want to be wary of moving too fast for people overall.

(To be clear, though, I'm not entirely sure which way I think is better.)

Reply:
> @Centril you've mentioned deadlines a few times - can you elaborate on where you find the perception of these come from? Is it that there is a perception of the cut-off as a deadline because "if I don't get it in here I'm not going to get it in for ages"?

I think that's partly the case. Mostly, however, I think it is the effect that deadlines have themselves. Once we decided that e.g. NLL should make it into Edition 2018 well in advance of the bugs and other issues actually having been fixed, I think that made us more inclined to stick to that deadline (due to social pressures, ...) whether or not those bugs were fixed (this in turn created problems like https://github.com/rust-lang/rust/issues/59159). Similar issues happened with async/await where we also set deadlines well in advance of the issues being resolved. Another issue that we rushed was the module system, which many last minute decisions being made.

> My concern with pushing out the cut-off is it unintuitively _increases_ the pressure to get it in next time, because there's now no regular schedule and "who knows when the next chance will come around".

Fair points! I do think this is a possibility. I've sorta mixed together two distinct issues here, one regarding cut-off, and one regarding setting deadlines for stabilizations. My primary concern here is the latter aspect...

> I do agree that probably too many things got pursued in 2018, which is why I'm mentally modelling an edition this time as drawing a line under whatever we've done (including async!) and bundling 'obviously required' breaking changes with it. A 'rallying point' doesn't necessarily mean we have to get a bunch of things done in pursuit of something - it can also mean just taking stock of what we've achieved.

...which is why I find this to be a particularly *helpful distinction* and a key point! I love the idea of focusing editions more on documenting to users what we have already stabilized (e.g. slice patterns, https://github.com/rust-lang/rust/pull/67712) rather than a push to stabilize more on a deadline (e.g. specialization, at least based on today's prospects, heh). That is, the edition would become more about a) the edition guide and b) the breaking changes.

> Let me just phrase that another way. If we make editions run on a schedule, the emphasis for big features should be to get enough of the pre-work done so that their **deprecations** are part of the edition, but not necessarily any new additions. One nice thing about this particular edition cycle is that I don't think we have a lot of big features, so we're going to be able to just focus on smalle changes.

Yes, I agree entirely. I think this strikes a healthy work balance and solidifies editions as migration and documentation polish.

Reply:
Nitpick: aren't slice patterns a general language change available in all editions with a new enough compiler?

It also happened in the 2018 edition that things that happened to come out around that time were marketed as being "new in 2018 edition!" even though they were just new in the latest compiler versions and also available in 2015. I think that we should _very much_ keep edition marketing to _only_ the things that you get by changing 2018 to 2021 in your cargo file. Not anything else at all.

Reply:
> Nitpick: aren't slice patterns a general language change available in all editions with a new enough compiler?

Yes, they are, but they could go under "taking stock" / enhancing the documentation of "what has been improved in the language since the last 3 years".

> I think that we should _very much_ keep edition marketing to _only_ the things that you get by changing 2018 to 2021 in your cargo file. Not anything else at all.

That's... going to be a [very short list that fits in 1/4 page](https://doc.rust-lang.org/stable/edition-guide/rust-2018/edition-changes.html) most likely. =D That writing is something you finish up in maybe 2 hours of work for a really polished version (beyond the work in the compiler itself).

Reply:
Short and sweet is good. Great in fact. It means we didn't need many changes, which means things mostly went right the first time.

Immediately after 2018 came out it was hard to find material that maintained the distinction between 1.30 vs 1.31 and 2015 vs 2018. Many people got confused about if they had to change edition to get ability X or Y. I just don't want any of that repeated.

Just to pick an example, slice patterns is given in the edition guide as being part of 2018 edition with a minimum rust version of 1.26, which is clearly nonsense because 2018 edition landed in 1.31.

Reply:
> Immediately after 2018 came out it was hard to find material that maintained the distinction between 1.30 vs 1.31 and 2015 vs 2018. Many people got confused about if they had to change edition to get ability X or Y. I just don't want any of that repeated.

Yeah I very clearly remember discussing this confusing with you and others at the time on Discord. ;) I believe we can do better this time around with the experience of what not to do from last time. However it's also possible that we would add to the confusion.